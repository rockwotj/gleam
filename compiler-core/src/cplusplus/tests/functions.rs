use crate::assert_cpp;

#[test]
fn exported_functions() {
    assert_cpp!(
        r#"
pub fn add(x, y) {
    x + y
}"#,
    );
}

#[test]
fn private_functions() {
    assert_cpp!(
        r#"
fn add(x, y) {
    x + y
}"#,
    );
}

#[test]
fn anonymous_functions() {
    assert_cpp!(
        r#"
fn squared(x) {
  let y = fn(z) { z * z }
  y(x)
}
    "#
    )
}

#[test]
fn captured_variables() {
    assert_cpp!(
        r#"
fn do_stuff(x, y) {
  let f = fn(z) { { x * y } + z }
  f(x + y)
}
    "#
    )
}

#[test]
fn multiple_module_functions() {
    assert_cpp!(
        r#"
fn add(x, y) {
  x + y
}
fn subtract(x, y) {
  add(x, 0 - y)
}
    "#
    )
}

#[test]
fn generic_function() {
    assert_cpp!(
        r#"
fn identity(x) {
  x
}
fn wrapped(x) {
  identity(x)
}
        "#
        )
}

#[test]
fn specialized() {
    assert_cpp!(
        r#"
fn identity(x) {
  x
}
fn add(a, b) {
  identity(a) + identity(b)
}
        "#
        )
}
