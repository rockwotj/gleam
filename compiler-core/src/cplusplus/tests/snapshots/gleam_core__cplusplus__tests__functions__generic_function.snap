---
source: compiler-core/src/cplusplus/tests/functions.rs
expression: "\nfn identity(x) {\n  x\n}\nfn wrapped(x) {\n  identity(x)\n}\n        "
---
#ifndef MY_PACKAGE_MY_MODULE_H_
#define MY_PACKAGE_MY_MODULE_H_

#include <gleam.h>

namespace my {
namespace module {

template <typename T$8>
T$8 identity(T$8 x);

template <typename T$10>
T$10 wrapped(T$10 x);

} // namespace my
} // namespace module

#endif // MY_PACKAGE_MY_MODULE_H_

---
#include "module.h"

namespace my {
namespace module {

template <typename T$8>
T$8 identity(T$8 x) {
  return x;
};

template <typename T$10>
T$10 wrapped(T$10 x) {
  return ::my::module::_private::identity<T$10>(x);
};

} // namespace my
} // namespace module

